karapace/backup/backends/v3/schema.py

```diff
@@ -0,0 +1,84 @@
+"""
+karapace
+
+Copyright (c) 2023 Aiven Ltd
+See LICENSE for details
+"""
+from dataclasses import field
+from karapace.avro_dataclasses.models import AvroModel
+from karapace.dataclasses import default_dataclass
+from typing import Optional, Tuple
+
+import datetime
+import enum
+import uuid
+
+
+@enum.unique
+class ChecksumAlgorithm(enum.Enum):
+    # Note: We have to have a default value that is unknown here, as otherwise there
+    # would be no purpose at all in representing this as an enum, due to how the Avro
+    # semantics work. If we introduce a new algorithm, an old version of Karapace will
+    # infer the default value of an enum field, so if we didn't have an unknown value,
+    # it would default to xxhash3, and just be broken without making sense.
+    unknown = "unknown"
+    xxhash3_64_be = "xxhash3_64_be"
+
+
+@default_dataclass
+class DataFile(AvroModel):
+    filename: str
+    partition: int = field(metadata={"type": "long"})
+    checksum: bytes
+    record_count: int = field(metadata={"type": "long"})
+
+    def __post_init__(self) -> None:
+        assert self.record_count >= 0
+        assert self.partition >= 0
+        assert self.checksum
+        assert self.filename
+
+
+@default_dataclass
+class Metadata(AvroModel):
+    version: int = field(metadata={"type": "int"})
+    tool_name: str
+    tool_version: str
+    started_at: datetime.datetime
+    finished_at: datetime.datetime
+    topic_name: str
+    topic_id: Optional[uuid.UUID]
+    partition_count: int = field(metadata={"type": "int"})
+    data_files: Tuple[DataFile, ...]
+    checksum_algorithm: ChecksumAlgorithm = ChecksumAlgorithm.unknown
+
+    def __post_init__(self) -> None:
+        assert len(self.data_files) == self.partition_count
+        assert self.finished_at >= self.started_at
+        assert self.partition_count == 1
+        assert self.version == 3
```

Comment:
I don't think we should eliminate `ChecksumAlgorithm.unknown` here, I'd rather do it as-is, as it will allow [inspecting](https://github.com/aiven/karapace/pull/616/files) a file with a future algorithm.

Reply:
It's currently handled here: https://github.com/aiven/karapace/blob/64aeebe43a3f7133859788c9b2c7f1d59054df71/karapace/backup/backends/v3/backend.py#L27

Reply:
I don't know how useful `assert self.started_at >= 0` would be, as we're interpreting as `datetime`, that puts upper and lower bounds on the values. `datetime.datetime(1970, 1, 1, 1, 0)` other than being represented as `0` is kind of random in this context?

FWIW, this test should give confidence the concrete values are correct: https://github.com/aiven/karapace/blob/64aeebe43a3f7133859788c9b2c7f1d59054df71/tests/integration/backup/test_v3_backup.py#L272